packages/components/CONTRIBUTING.md

```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+	return <Button variant="primary" as="a" href={href}>{ children }</Button>;
+}
+```
+
+### Composition patterns
+
+TBD — E.g. Using `children` vs custom render props vs arbitrary "data" props
+
+### (Semi-)Controlled components
+
+TBD
+
+### Layout "responsibilities"
+
+TBD — Components' layout responsibilities and boundaries (i.e., a component should only affect the layout of its children, not its own) -->
+
+#### Components & Hooks
+
+One way to enable reusability and composition is to extract a component's underlying logic into a hook (living in a separate `hook.ts` file). The actual component (usually defined in a `component.tsx` file) can then invoke the hook and use its output to render the required DOM elements. For example:
+
+```tsx
+// in `hook.ts`
+function useExampleComponent( props: PolymorphicComponentProps< ExampleProps, 'div' > ) {
+	// Merge received props with the context system.
+	const { isVisible, className, ...otherProps	} = useContextSystem( props, 'Example' );
+
+	// Any other reusable rendering logic (e.g. computing className, state, event listeners...)
+	const cx = useCx();
+	const classes = useMemo(
+		() =>
+			cx(
+				styles.example,
+				isVisible && styles.visible,
+				className
+			),
+		[ className, isVisible ]
+	);
+
+	return {
+		...otherProps,
+		className: classes
+	};
+}
+
+// in `component.tsx`
+function Example(
+	props: PolymorphicComponentProps< ExampleProps, 'div' >,
+	forwardedRef: Ref< any >
+) {
+	const exampleProps = useExampleComponent( props );
+
+	return <View { ...spacerProps } ref={ forwardedRef } />;
+}
+```
+
+A couple of good examples of how hooks are used for composition are:
+
+- the `Card` component, which builds on top of the `Surface` component by [calling the `useSurface` hook inside its own hook](/packages/components/src/card/card/hook.js);
+- the `HStack` component, which builds on top of the `Flex` component and [calls the `useFlex` hook inside its own hook](/packages/components/src/h-stack/hook.js).
+
+<!-- ### APIs Consinstency
+
+[To be expanded] E.g.:
+
+- Boolean component props should be prefixed with `is*` (e.g. `isChecked`), `has*` (e.g. `hasValue`) or `enable*` (e.g. `enableScroll`)
+- Event callback props should be prefixed with `on*` (e.g. `onChanged`)
+- Subcomponents naming conventions (e.g `CardBody` instead of `Card.Body`)
+- ...
+
+### Performance
+
+TDB -->
+
+### Technical requirements for new components
+
+The following are a set of technical requirements for all newly introduced components. These requirements are also retroactively being applied to existing components.
+
+For an example of a component that follows these requirements, take a look at [`ItemGroup`](/packages/components/src/item-group) as an example.
+
+#### TypeScript
+
+We strongly encourage using TypeScript for all new components. Components should be typed using the `WordPressComponent` type (more details about polymorphism can be found above in the "Components composition" section).
```

Comment:
There is no "Components composition" section yet so the part in parenthesis should be commented out for now.

Reply:
Done in [`6647446` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/66474469b39466210d2596af1110719ea4e84602)